components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java

```diff
@@ -747,4 +749,76 @@ private boolean isFormatCorrect(String regularExpression, String attribute) {
+     * @throws IdentityEventException
+     * @throws IdentityRecoveryException
+     */
+    public void preValidatePasswordWithConfirmationKey(String confirmationKey, String password) throws
+            IdentityEventException, IdentityRecoveryException {
+
+        UserRecoveryDataStore userRecoveryDataStore = JDBCRecoveryDataStore.getInstance();
+        UserRecoveryData recoveryData = userRecoveryDataStore.load(confirmationKey);
+        User user = recoveryData.getUser();
+        String userStoreDomain = user.getUserStoreDomain();
+        String username = user.getUserName();
+
+        preValidatePassword(username, password, userStoreDomain);
+    }
+
+    /**
+     * Pre validate the password against the policies defined in the Identity Server.
+     *
+     * @param username Username
+     * @param password Password to be pre-validated against the policies defined in IS.
+     * @throws IdentityRecoveryServerException
+     * @throws IdentityEventException
+     */
+    public void preValidatePasswordWithUsername(String username, String password) throws IdentityEventException,
+            IdentityRecoveryServerException {
+
+        String userStoreDomain = IdentityUtil.extractDomainFromName(username);
+        username = UserCoreUtil.removeDomainFromName(username);
+
+        preValidatePassword(username, password, userStoreDomain);
+    }
+
+    private void preValidatePassword(String username, String password, String userStoreDomain) throws
+            IdentityRecoveryServerException, IdentityEventException {
+
+        String tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+        String eventName = IdentityEventConstants.Event.PRE_UPDATE_CREDENTIAL_BY_ADMIN;
+        HashMap<String, Object> properties = new HashMap<>();
+
+        properties.put(IdentityEventConstants.EventProperty.USER_NAME, username);
+        properties.put(IdentityEventConstants.EventProperty.CREDENTIAL, password);
+        properties.put(IdentityEventConstants.EventProperty.TENANT_DOMAIN, tenantDomain);
+        properties.put(IdentityEventConstants.EventProperty.TENANT_ID, tenantId);
+
+        RealmService realmService = IdentityRecoveryServiceDataHolder.getInstance().getRealmService();
+        try {
+            AbstractUserStoreManager userStoreManager = (AbstractUserStoreManager)
+                    realmService.getTenantUserRealm(tenantId).getUserStoreManager();
+            UserStoreManager secondaryUserStoreManager = userStoreManager.getSecondaryUserStoreManager
+                    (userStoreDomain);
+            properties.put(IdentityEventConstants.EventProperty.USER_STORE_MANAGER, secondaryUserStoreManager);
+        } catch (UserStoreException e) {
+            throw Utils.handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED, username,
```

Comment:
For the Error `IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED` can you append error details?

Reply:
Shall we add the debug log here? WDYT?

Reply:
fixed in https://github.com/wso2-extensions/identity-governance/pull/355